keps/sig-node/4381-dra-structured-parameters/README.md

```diff
@@ -557,278 +532,231 @@ reconstructed by the driver. This has no effect on already allocated claims
+blocks.
 
 ```yaml
-kind: CardParameters
-apiVersion: dra.example.com/v1alpha1
-metadata:
-  name: my-parameters
-  namespace: user-namespace
-  uid: foobar-uid
+kind: ResourceSlice
+apiVersion: resource.k8s.io/v1alpha3
 ...
 spec:
-  minimumRuntimeVersion: v12.0.0
-  minimumMemory: 32Gi
-  # "sharing" is a configuration parameter that does not
-  # get translated into the selector below.
-  sharing:
-    strategy: TimeSliced
-```
-
-Note that all fields in this CRD can be fully validated since it is owned by
-the DRA driver itself. This includes value ranges that are specific to the
-underlying hardware. There's no risk of using invalid attribute names because
-only the fields shown here are valid.
-
-With this CRD in place, a DRA driver controller is able to convert instances of
-it into a generic, in-tree `ResourceClaimParameters` object that the scheduler
-is able to understand.
-
-For the example above, the converted object would look as follows:
-
-```yaml
-kind: ResourceClaimParameters
-apiVersion: resource.k8s.io/v1alpha2
-
-metadata:
-  # This cannot be the same as my-parameters because parameter objects with a different
-  # type might also use it. Instead, the original object gets linked to below.
-  name: someArbitraryName
-  namespace: user-namespace
-
-generatedFrom:
-  name: my-parameters
-  kind: CardParameters
-  apiGroup: dra.example.com
-  uid: foobar-uid
-
-vendorParameters:
-  # A vendor can put any kind of object here to pass the configuration
-  # parameters down to the kubelet plugin. In this case, the vendor
-  # driver simply copied the entire CR. It could also be some
-  # separate, smaller configuration type.
-  #
-  # Beware that ResourceClaimParameters have separate RBAC rules than
-  # the vendor CRD, so information included here may get visible
-  # to more users than the original CRD. Both objects are in the same
-  # namespace.
-  kind: CardParameters
-  apiVersion: dra.example.com/v1alpha1
-  metadata:
-    name: my-parameters
-    namespace: user-namespace
-    uid: foobar-uid
-  ...
-  spec:
-    minimumRuntimeVersion: v12.0.0
-    minimumMemory: 32Gi
-    sharing:
-      strategy: TimeSliced
-
-driverRequests:
-- driverName: cards.dra.example.com
-  requests:
-  # Each entry here is a request for one resource.
-  - namedresources:
-      # Selectors are CEL expressions with access to the attributes of the named resource
-      # that is being checked for a match.
-      selector: |-
-        attributes.version["runtimeVersion"].isGreaterThan(semver("12.0.0")) &&
-        attributes.quantity["memory"].isGreaterThan(quantity("32Gi"))
+  # The node name indicates the node.
+  # Each driver on a node provides pools of devices for allocation,
+  # with unique device names inside each pool.
+  # Usually, but not necessarily, that pool name is the same as the
+  # node name.
+  nodeName: worker-1
+  poolName: worker-1
+  driverName: gpu.dra.example.com
+  sharedCapacity:
+  - name: gpu-0-memory-block-0
```

Comment:
What if big gpus with say ~100 slices come in the future? Do we have to then explicitly list all possible partitions? The object size will become too big.

Reply:
https://github.com/kubernetes/enhancements/pull/4709#issuecomment-2164181005

Reply:
Oh gross.  I just realized this is describing shared resources as per-slice and then jamming it into the existing device struct.  So yeah, if you have a machine with 8 GPUs and each GPU has 16 memory blocks, you end up with 8*16 = 128 items with the exact same names but for the indices, and the exact same capacities, plus 31 device definitions (assuming naturally aligned partitions), all with mostly the same attributes.  Is that right? 

I had to write a program to visualize it: https://go.dev/play/p/KY_4x4tOJzk

With just about the most reductive naming I can imagine and NO OTHER attributes besides memory, this is 50K of boilerplate.  Expanding the names to match this example makes it 60K.  Adding a couple of attributes (literally 2) makes it 80K.

These are not huge numbers on their own BUT I think we agree that attributes are likely to be more than just 3, and the multiplier effect is pretty large (every attribute is repeated ~250 times in this example).

I am afraid that this will be a sticking point for future progress.  Pulling attributes to a per-pool common list is one step.  I am open to other ideas for how to scale down the redundant repetitive duplicated information.

Aside from that, I recall a thread on the evolution prototype about whether something like:

```
ResourceSlice:
  devices:
  - name: aaa
  - name: bbb
  - name: ccc
```

where device was a simple struct, and 

```
ResourceSlice:
  partitionableDevices:
  - name: aaa
    partitions:
    - name: aaa-1
    - name: aaa-2
    - name: aaa-3
  - name: bbb
    partitions:
    - name: bbb-1
    - name: bbb-2
    - name: bbb-3
  - name: ccc
    partitions:
    - name: ccc-1
    - name: ccc-2
    - name: ccc-3
```

or

```
ResourceSlice:
  devices:
  - partitioned:
      name: aaa
      partitions:
      - name: aaa-1
      - name: aaa-2
      - name: aaa-3
  - partitioned:
    name: bbb
      partitions:
      - name: bbb-1
      - name: bbb-2
      - name: bbb-3
  - partitioned:
    name: ccc
      partitions:
      - name: ccc-1
      - name: ccc-2
      - name: ccc-3
```

where `partitionableDevice` was a more complex struct, which left room for other "forms" of devices.  What happened to that discussion?  The advantage here being that the `sharedCapacity` would be per "partitionableDevice" not per pool, making the names shorter automatically (though not less redundant).

Reply:
Even if we were to push `partitions` into the device like this (which I'm not a fan of), we still end up with the same amount of redundant code (as you mentioned). 

If the only concern is to make names shorter, I would prefer to do that like I proposed here https://github.com/kubernetes-sigs/wg-device-management/pull/27#discussion_r1634768662. Because this supports pulling from both on-chip and off-chip shared resources to define a device (if available). From my perspective its also simpler to parse / reason than nested devices would be.

i.e.
```
apiVersion: resource.k8s.io/v1alpha3
kind: ResourceSlice
metadata:
  name: <uuid>
spec:
  driver: gpu.nvidia.com
  poolName: node0
  nodeName: node0
  sharedCapacity:
  - name: gpu-0-shared-capacity
    resources: 
    - name: memory
      capacity: 40Gi
    - name: compute
      capacity: 180
  - name: gpu-1-shared-capacity
    resources: 
    - name: memory
      capacity: 40Gi
    - name: compute
      capacity: 180
  devices:
  - name: gpu-0
    sharedCapacityConsumed:
    - name: gpu-0-shared-capacity
      resources: 
      - name: memory
        capacity: 40Gi
      - name: compute
        capacity: 180
  - name: gpu-0-first-half
    sharedCapacityConsumed:
    - name: gpu-0-shared-capacity
      resources: 
      - name: memory
        capacity: 20Gi
      - name: compute
        capacity: 90
  - name: gpu-0-second-half
    sharedCapacityConsumed:
    - name: gpu-0-shared-capacity
      resources: 
      - name: memory
        capacity: 20Gi
      - name: compute
        capacity: 90
  - name: gpu-1
    sharedCapacityConsumed:
    - name: gpu-1-shared-capacity
      resources: 
      - name: memory
        capacity: 40Gi
      - name: compute
        capacity: 180
  - name: gpu-1-first-half
    sharedCapacityConsumed:
    - name: gpu-1-shared-capacity
      resources: 
      - name: memory
        capacity: 20Gi
      - name: compute
        capacity: 90
  - name: gpu-1-second-half
    sharedCapacityConsumed:
    - name: gpu-1-shared-capacity
      resources: 
      - name: memory
        capacity: 20Gi
      - name: compute
        capacity: 90
```

Reply:
Also, in terms of memory blocks specifically, this is another argument for introducing something like my proposed  [`intRange` type](https://github.com/kubernetes-sigs/wg-device-management/pull/6/files#diff-bf61dad125edb47c64e4ea2df7f4690a79e4588b3decdec1b97c603870270db0).

Each memory block range can then be represented as a range (or comma separated list of non-overlapping ranges) of integers, rather than requiring one entry per block.

i.e.:
```yaml
kind: ResourceSlice
apiVersion: resource.k8s.io/v1alpha3
  # The node name indicates the node.
  # Each driver on a node provides pools of devices for allocation,
  # with unique device names inside each pool.
  # Usually, but not necessarily, that pool name is the same as the
  # node name.
  nodeName: worker-1
  poolName: worker-1
  driverName: gpu.dra.example.com
  sharedCapacity:
  - name: gpu-0-memory-blocks
    intRange: "0-3"
 devices:
  - name: gpu-0
    attributes:
    - name: memory
      quantity: 40Gi
    sharedCapacityConsumed:
    - name: gpu-0-memory-blocks
      intRange: "0-3"
  - name: gpu-0-first-half
    attributes:
    - name: memory
      intRange: 20Gi
    sharedCapacityConsumed:
    - name: gpu-0-memory-blocks
      intRange: "0-1"
  - name: gpu-0-second-half
    attributes:
    - name: memory
      quantity: 20Gi
    sharedCapacityConsumed:
    - name: gpu-0-memory-blocks
      intRange: "2-3"
  - name: gpu-0-middle-half
    attributes:
    - name: memory
      quantity: 20Gi
    sharedCapacityConsumed:
    - name: gpu-0-memory-blocks
      intRange: "1-2"
 - name: gpu-0-first-quarter
    attributes:
    - name: memory
      intRange: 10Gi
    sharedCapacityConsumed:
    - name: gpu-0-memory-blocks
      intRange: "0"
 - name: gpu-0-second-quarter
    attributes:
    - name: memory
      quantity: 10Gi
    sharedCapacityConsumed:
    - name: gpu-0-memory-blocks
      intRange: "1"
 - name: gpu-0-third-quarter
    attributes:
    - name: memory
      quantity: 10Gi
    sharedCapacityConsumed:
    - name: gpu-0-memory-blocks
      intRange: "2"
 - name: gpu-0-fourth-quarter
    attributes:
    - name: memory
      quantity: 10Gi
    sharedCapacityConsumed:
    - name: gpu-0-memory-blocks
      intRange: "3"
```

Reply:
It doesn't. That was just an example of a hypothetical device that didn't care about alignment (just that the scheduler should have the capability to allocate either the full device, or two halves of a device). Because of the way the scheduler component of this works, once it decided to allocate a half of the device, only the other half would be available (because not enough of the shared memory / compute resources would be available to allow the full device to be allocated). It was purposefully simplified for illustration purposes.

What you are describe: "a device that supports any power of two size, as long as it is aligned on a multiple of the size", would need to be modeled differently. We don't have a compact way to represent that yet, but I could imagine adding one a new field to the one-of (i.e. the one with either a `Quantity` or a `IntRange) which could represent that compactly.

Reply:
Take a look at https://github.com/kubernetes-sigs/wg-device-management/issues/20#issuecomment-2168189769 - in particular Option 4. I think it:
1) Is quite understandable
2) in the non-partitioned case reduces to "common attributes"
3) reduces boilerplate by a factor of 10

Int-range would shrink it a bit more.